Classes/Emogrifier.php

```diff
@@ -1289,41 +1292,71 @@ function (array $matches) {
         $trimmedLowercaseSelector = trim($lowercasePaddedSelector);
         $xPathKey = md5($trimmedLowercaseSelector);
         if (!isset($this->caches[self::CACHE_KEY_XPATH][$xPathKey])) {
-            $roughXpath = '//' . preg_replace(
-                array_keys($this->xPathRules),
-                $this->xPathRules,
-                $trimmedLowercaseSelector
-            );
-            $xPathWithIdAttributeMatchers = preg_replace_callback(
-                self::ID_ATTRIBUTE_MATCHER,
-                [$this, 'matchIdAttributes'],
-                $roughXpath
-            );
-            $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
-                self::CLASS_ATTRIBUTE_MATCHER,
-                [$this, 'matchClassAttributes'],
-                $xPathWithIdAttributeMatchers
-            );
+            $notSelector = null;
+            if (strstr($trimmedLowercaseSelector, ':not')) {
+                list($trimmedLowercaseSelector, $notSelector) = explode(':not', $trimmedLowercaseSelector);
+            }
 
-            // Advanced selectors are going to require a bit more advanced emogrification.
-            // When we required PHP 5.3, we could do this with closures.
-            $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
-                '/([^\\/]+):nth-child\\(\\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthChild'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
-            $finalXpath = preg_replace_callback(
-                '/([^\\/]+):nth-of-type\\(\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthOfType'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
+            $finalXpath = '//' . $this->translateCssToXpathPass($trimmedLowercaseSelector);
+
+            if ($notSelector) {
+                $notSelector = trim($notSelector, '()');
+                $finalXpath = $finalXpath . '[not(' . $this->translateCssToXpathPass($notSelector, true) . ')]';
+            }
 
             $this->caches[self::CACHE_KEY_SELECTOR][$xPathKey] = $finalXpath;
         }
         return $this->caches[self::CACHE_KEY_SELECTOR][$xPathKey];
     }
 
     /**
+     * @param string $trimmedLowercaseSelector
+     *
+     * @param bool $inline If set to true it only outputs inline part of xPath class selector. Needed for :not selector
```

Comment:
No behavior toggles as parameter, please. Instead, please let's have two separate methods that each do exactly one thing (while one method may call the other).

Reply:
You are welcome to change that. I'm not going to duplicate code - that's much worse solution.

Reply:
I think you misunderstood me. I'll create a PR for this refactoring.